Semantics

Hixie opened the pull request:
The following open issues exist:
- Sliders, drop-down buttons, and text input fields aren't accessible yet.
- Time picker isn't accessible yet. This will require making it possible to create custom SemanticsNode subtrees.
- The Java/Objective C sides here aren't done.

See https://github.com/flutter/engine/pull/2277 for the mojom used by this patch.

Hixie committed:
Semantics

Hixie commented on examples/stocks/lib/main.dart:
```diff
@@ -104,17 +104,19 @@ class StocksAppState extends State<StocksApp> {
-      },
-      onGenerateRoute: _getRoute,
-      onLocaleChanged: _onLocaleChanged
+    return new SemanticsDebugger(
```
before checkin I will remove the SemanticsDebuggers around this app and the earlier one. Later I'll add a way to enable it from MaterialApp.

Hixie commented on examples/stocks/lib/stock_home.dart:
```diff
@@ -88,16 +88,25 @@ class StockHomeState extends State<StockHome> {
+                        new Container(
+                          width: 8.0
+                        ),
+                        new Text('DUMP SEMANTICS'),
```
I'll revert this change.

Hixie commented on packages/flutter/lib/src/material/radio.dart:
```diff
@@ -100,7 +103,9 @@ class _RenderRadio extends RenderToggleable {
   );
 
-  bool get isInteractive => super.isInteractive && !value;
+  bool get isInteractive {
```
i can revert this, it was for being able to add print statements for debugging

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,309 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
 RenderingExceptionHandler debugRenderingExceptionHandler;
 
+
+// TODO(ianh): factor out common parts in the classes below -- DO NOT CHECKIN IN CURRENT STATE
```
this comment is out of date, I've done the refactoring.

Hixie commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,446 @@
+            // ancestors. In that case, we drop the child eagerly here.
+            // TODO(ianh): Find a way to assert that the same node didn't
+            // actually appear in the tree in two places.
+            child.parent?.dropChild(child);
```
I really don't like that we're building the tree upside down. It'd be much cleaner if it built it straight down. I think this might be part of the cause of the bug where things are getting lost (drop the popup menu, cancel it, drop it again, the box that would be getting the cancel isn't included the second time).

Hixie commented on packages/flutter/lib/src/widgets/gesture_detector.dart:
```diff
@@ -309,3 +343,122 @@ class _GestureDetectorState extends State<GestureDetector> {
   }
 }
+
+class GestureSemantics extends OneChildRenderObjectWidget {
```
maybe this should be private

Hixie commented:
cc @abarth

abarth commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -87,24 +87,27 @@ class SnackBar extends StatelessComponent {
-                  children: children,
-                  alignItems: FlexAlignItems.center
+        child: new Semantics(
+          container: true,
```
I wonder if this belongs in Material somehow.

Hixie replied:
I think we use Material in too many places. container:true is relatively heavy-duty, I don't think we want to make it automatic when you insert a new material layer. Might be worth looking at in more detail in a subsequent patch, though.

abarth commented on packages/flutter/lib/src/material/toggleable.dart:
```diff
@@ -79,9 +81,20 @@ abstract class RenderToggleable extends RenderConstrainedBox {
+    final bool wasInteractive = isInteractive;
+    _onChanged = value;
+    if (wasInteractive != isInteractive) {
+      markNeedsPaint();
```
I guess this was a pre-existing bug.

abarth commented on packages/flutter/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -112,13 +112,16 @@ class RenderEditableParagraph extends RenderParagraph {
     }
   }
 
+  bool get _hasVisualOverflow => (_contentSize.width > size.width);
```
No need for parens

Hixie replied:
fixed

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,309 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+  _SemanticsGeometry() : transform = new Matrix4.identity();
+  _SemanticsGeometry.withClipFrom(_SemanticsGeometry other) {
+    if (other != null)
+      clipRect = other.clipRect;
```
Elvis operator: `?.`

Hixie replied:
done

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,309 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+      _max4(point1.x, point2.x, point3.x, point4.x),
+      _max4(point1.y, point2.y, point3.y, point4.y)
+    );
+  }
```
This function should probably be in MatrixUtils.  SkMatrix has it, for example.

Hixie replied:
done

Hixie commented:
@abarth per your request:
https://github.com/Hixie/flutter/blob/semantics-3/packages/flutter/lib/src/rendering/README.md
Let me know if there's specific questions this should answer that it doesn't answer. I think the key part is the part I've marked in bold. Once I realised that (it was the last sentence I wrote in the document), the rest seemed kinda redundant and I wasn't sure how much detail to put into it after that.

Hixie commented:
I plan to add more tests but modulo that this is ready for a more serious review.

abarth commented on packages/flutter/pubspec.yaml:
```diff
@@ -9,7 +9,9 @@ dependencies:
-  sky_services: 0.0.87
+#  sky_services: 0.0.87
+  sky_services:
+    path: ../../../engine/src/out/Debug/gen/dart-pkg/sky_services
```
Obviously not for landing :)

abarth commented on packages/flutter/lib/src/widgets/semantics_debugger.dart:
```diff
@@ -0,0 +1,346 @@
+  Point _lastTapLocation;
+  void _handlePointerDown(PointerDownEvent event) {
+    setState(() {
+      _lastTapLocation = event.position;
```
Why not use `onTapDown`?  That should give you the down position for the tap.

Hixie replied:
Ah, yeah. Originally this was done with onTapDown but we don't always get that (e.g. for a drag if it's quick enough you don't get a tapdown). I should indeed rename the variable though.

Hixie replied:
done.

abarth commented on packages/flutter/lib/src/widgets/semantics_debugger.dart:
```diff
@@ -0,0 +1,346 @@
+    }
+    message = message.trim();
+    if (message != '') {
+      textPainter = new TextPainter(new StyledTextSpan(textStyles, <TextSpan>[new PlainTextSpan(message)]));
```
Why not use ??= for the allocation and then use the `text=` setter?

Hixie replied:
I've had too much exposure to the Widgets patterns? :-P

Good call, will fix.

abarth commented on packages/flutter/lib/src/widgets/semantics_debugger.dart:
```diff
@@ -0,0 +1,346 @@
+import 'framework.dart';
+import 'gesture_detector.dart';
+
+class SemanticsDebugger extends StatefulComponent {
```
The debugger really is like a whole mini copy of the entire system.

abarth commented on packages/flutter/lib/src/widgets/modal_barrier.dart:
```diff
@@ -24,17 +25,20 @@ class ModalBarrier extends StatelessComponent {
-          decoration: new BoxDecoration(
-            backgroundColor: color
+    return new Semantics(
+      container: true,
```
This is a bit mysterious to me currently.  What does the container annotation mean and how can I remember when to use it?

Hixie replied:
Added dartdocs for this.

abarth commented on packages/flutter/lib/src/widgets/gesture_detector.dart:
```diff
@@ -31,6 +31,13 @@ export 'package:flutter/gestures.dart' show
+/// onLongPress, a vertical drag in either onVerticalDragUpdate or
+/// onPanUpdate, and a horizontal drag in either
+/// onHorizontalDragUpdate or onPanUpdate. The corresponding down/up
+/// or start/end events do _not_ fire in this case.
```
That seems slightly dangerous.  Can the existing clients actually cope with that?

Hixie replied:
No, they can't, which is why I changed it as you noted. Will remove this comment.

abarth commented on packages/flutter/lib/src/widgets/gesture_detector.dart:
```diff
@@ -309,3 +343,122 @@ class _GestureDetectorState extends State<GestureDetector> {
+      if (onVerticalDragUpdate != null)
+        onVerticalDragUpdate(delta);
+      if (onVerticalDragEnd != null)
+        onVerticalDragEnd(Offset.zero);
```
Looks like you do actually fire the start/end parts.  That seems easier for consumers.

abarth commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -2009,21 +2009,73 @@ class RepaintBoundary extends OneChildRenderObjectWidget {
+
+  final bool container;
+  final bool checked;
+  final String label;
```
dartdoc comments would be helpful here.

Hixie replied:
done for the three semantics classes in basic.dart

abarth commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -2009,21 +2009,73 @@ class RepaintBoundary extends OneChildRenderObjectWidget {
+  void updateRenderObject(RenderSemanticAnnotations renderObject, Semantics oldWidget) {
+    renderObject.container = container;
+    renderObject.checked = checked;
+    renderObject.label = label;
```
We should probably use `..` cascade here.  I realize you're following the existing pattern, but this just occurred to me.

Hixie replied:
Filed https://github.com/flutter/flutter/issues/1381. Not sure it's worth it, though. I would hope the resulting code would be identical, and with the current pattern the code is more consistent.

abarth commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -2009,21 +2009,73 @@ class RepaintBoundary extends OneChildRenderObjectWidget {
+
+class MergeSemantics extends OneChildRenderObjectWidget {
+  MergeSemantics({ Key key, Widget child }) : super(key: key, child: child);
+  RenderSemanticMerger createRenderObject() => new RenderSemanticMerger();
```
Why the name change?

Hixie replied:
fixed

abarth commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -2009,21 +2009,73 @@ class RepaintBoundary extends OneChildRenderObjectWidget {
+
+class ExcludeSemantics extends OneChildRenderObjectWidget {
+  ExcludeSemantics({ Key key, Widget child }) : super(key: key, child: child);
+  RenderSemanticExcluder createRenderObject() => new RenderSemanticExcluder();
```
ditto

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+import 'dart:ui' show Rect;
+
+import 'package:flutter/painting.dart';
+import 'package:sky_services/semantics/semantics.mojom.dart' as engine;
```
We usually import these as `mojom`

Hixie replied:
fixed

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+
+import 'node.dart';
+
+typedef void SemanticAnnotator(SemanticsNode semantics);
```
Good place for some dartdoc

Hixie replied:
added.

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+
+/// A BitField over an enum (or other class whose values implement "index").
+/// Only the first 63 values of the enum can be used as indices.
+class BitField<T extends dynamic> {
```
Please move this class into its own file.  We should probably move it into another package, but for now a separate file is fine.

Hixie replied:
moved to basic_types for now

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+  // These are automatically handled by RenderObject's own logic
+
+  Matrix4 get transform => _transform;
+  Matrix4 _transform = new Matrix4.identity();
```
I wish we had an immutable matrix.  Maybe we should continue pretending that Matrix4 is immutable and have a common static final that we use as the initializer here.  Otherwise, we're churning memory for no real gain.

Hixie replied:
To avoid churn even further, I've made all the code handle "null" as meaning "identity matrix". This avoids us accidentally mutating our "identity" static.

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+
+  BitField<_SemanticFlags> _flags = new BitField<_SemanticFlags>.filled(_SemanticFlags.values.length, false);
+
+  void _flagSetter(_SemanticFlags flag, bool value, { bool needsHandler: false }) {
```
_setFlag

Hixie replied:
done

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+
+  List<SemanticsNode> _children;
+  bool get hasChildren => _children?.isNotEmpty ?? false;
+  bool _needsCleanup = false;
```
I usually call this something like `_dead`

Hixie replied:
done

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+    List<SemanticsNode> oldChildren = _children;
+    _children = _newChildren;
+    _newChildren = oldChildren;
+    oldChildren?.clear();
```
nit: I'd clear the list before assigning it `_newChildren` to avoid the WTF factor when reading the lines of code in order.

Hixie replied:
oh yeah, that's better. fixed.

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+  }
+  void detach() {
+    super.detach();
+    assert(_nodes.containsKey(_id));
```
assert(! _detachedNodes.contains(this))

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+    if (_dirty)
+      return;
+    _dirty = true;
+    _dirtyNodes.add(this);
```
assert(! _dirtyNodes.contains(this))

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+    _dirtyNodes.sort((SemanticsNode a, SemanticsNode b) => a.depth - b.depth);
+    for (int index = 0; index < _dirtyNodes.length; index += 1) {
+      // we mutate the list as we walk it here, which is why we use an index instead of an iterator
+      SemanticsNode node = _dirtyNodes[index];
```
assert(node._dirty)

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+    _dirtyNodes.sort((SemanticsNode a, SemanticsNode b) => a.depth - b.depth);
+    List<engine.SemanticsNode> updatedNodes = <engine.SemanticsNode>[];
+    for (SemanticsNode node in _dirtyNodes) {
+      assert(node.parent?._dirty != true);
```
assert(node._dirty)

abarth commented on packages/flutter/lib/src/rendering/semantics.dart:
```diff
@@ -0,0 +1,456 @@
+    List<engine.SemanticsNode> updatedNodes = <engine.SemanticsNode>[];
+    for (SemanticsNode node in _dirtyNodes) {
+      assert(node.parent?._dirty != true);
+      if (node._dirty)
```
How can the node not be dirty if its in the dirty list?

Hixie replied:
I'll add a comment to explain it in the code, but for now: `_serialize()` does a deep serialisation of all contiguous dirty nodes. So it can end up cleaning more than just the top node of the serialisation.

abarth commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -642,15 +660,20 @@ abstract class _RenderCustomClip<T> extends RenderProxyBox {
   }
 
   T get _defaultClip;
+  Rect get _defaultApproximateClipRect => Point.origin & size;
```
I'd just inline this into the one caller.

Hixie replied:
done

abarth commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1336,3 +1403,202 @@ class RenderMetaData extends RenderProxyBox {
+    GestureLongPressCallback onLongPress,
+    GestureDragUpdateCallback onHorizontalDragUpdate,
+    GestureDragUpdateCallback onVerticalDragUpdate,
+    this.scrollFactor: 0.8
```
0.8?

Hixie replied:
So that when you scroll you still see some of the previous page. This is completely arbitrary though. Would you prefer 1.0? I expect to tweak this based on real user feedback eventually.

abarth commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1336,3 +1403,202 @@ class RenderMetaData extends RenderProxyBox {
+  void set onTap(GestureTapCallback value) {
+    if (_onTap == value)
+      return;
+    bool hadSemantics = hasSemantics;
```
had and has are pretty similar.  Easy to mis-read.  Maybe didHaveSemantics?

Hixie replied:
good idea. fixed.

abarth commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1336,3 +1403,202 @@ class RenderMetaData extends RenderProxyBox {
+    assert(container != null);
+  }
+
+  bool get container => _container;
```
dartdoc for these would be welcome.  can be the same as the widget layer.

Hixie replied:
added

abarth commented on packages/flutter/lib/src/rendering/overflow.dart:
```diff
@@ -236,4 +236,5 @@ class RenderOffStage extends RenderBox with RenderObjectWithChildMixin<RenderBox
 
   bool hitTest(HitTestResult result, { Point position }) => false;
   void paint(PaintingContext context, Offset offset) { }
+  void visitChildrenForSemantics(RenderObjectVisitor visitor) { }
```
Should this default to just calling the normal visitChildren?

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+    return clipRect.intersect(other);
+  }
+  Matrix4 transform;
+  void applyAncestorChain(List<RenderObject> ancestorChain) {
```
This looks like an antipattern.  Generally these values are much cheaper to compute as you walk down the tree.  You generally keep a clip and transform stack as you walk down so you don't so much redundant work.

Hixie replied:
In theory these's no redundant work here. For each parent-child step, we only ever compute it once. When there's multiple children, we fork the matrix (see the _ForkingSemanticsFragment class).

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+      parent.applyPaintTransform(child, transform);
+    }
+  }
+  void updateSemanticsNode({ RenderObject rendering, SemanticsNode semantics, SemanticsNode parentSemantics }) {
```
I wonder if we can extract a SemanticsProvider interface from RenderObject that we can use in these classes instead of RenderObject.  That might let us move this code out of object.dart into semantics.dart.

Hixie replied:
Interesting idea.

Hixie replied:
Filed https://github.com/flutter/flutter/issues/1383 to track this.

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+      semantics.wasAffectedByClip = true;
+    } else {
+      semantics.rect = rendering.semanticBounds;
+      semantics.wasAffectedByClip = parentSemantics?.wasAffectedByClip ?? false;
```
This looks like a nice downward walk that could compute the clips.

Hixie replied:
This is always called immediately after `geometry.applyAncestorChain()`. It's the same walk. It's just skipped when the node doesn't correspond to one with actual semantics.

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+    assert(currentSemantics == null);
+    assert(parentSemantics == null);
+    owner._semantics ??= new SemanticsNode.root(
+      handler: owner is SemanticActionHandler ? owner as dynamic : null
```
Hello dynamic_cast, I've missed you so.

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
+    geometry.applyAncestorChain(_ancestorChain);
+    for (_SemanticsFragment child in _children) {
+      assert(child._ancestorChain.last == owner);
+      yield* child.compile(
```
neat

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+    }
+    if (!noGeometry && (_semantics == null || (_semantics.hasChildren && _semantics.wasAffectedByClip))) {
+      // Since the geometry might have changed, we need to make sure to reapply any clips.
+      visitChildrenForSemantics((RenderObject child) {
```
Woah there.  Why do we need to walk our children here?  Can't we just set a dirty bit and visit the children later?

Hixie replied:
Oh, yeah, that would work. Set a super-dirty bit and then when walking the subtree later, if the bit is set, pass the bit down to make sure the child ignores its dirtyness.

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+        node = node.parent;
+        node._needsSemanticsUpdate = true;
+        node._semantics?.reset();
+      } while (node._needsSemanticsUpdate || node._semantics == null || foundSemanticsNodes < 3);
```
We just set _needsSemanticsUpdate to true on line 1531.  How does this loop ever terminate?

Also, 3.

Hixie replied:
This loop is totally busted. I'll have another go at it.

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+    assert(fragment is _InterestingSemanticsFragment);
+    SemanticsNode node = fragment.compile(parentSemantics: _semantics?.parent).single;
+    assert(node != null);
+    assert(node == _semantics);
```
magic

Hixie replied:
Yeah. Unfortunately I need it updated through the tree, not just at the top, so it wasn't clear how to do this without magic. (That's why I added the assert.)

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+    assert(node == _semantics);
+  }
+
+  _SemanticsFragment _getSemantics() {
```
_getSemanticsFragment?

Hixie replied:
done

abarth commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+  /// in the case where only the labels or flags changed. If you pass
+  /// 'noGeometry: true' when the geometry did change, the semantic
+  /// tree will be out of date.
+  void markNeedsSemanticsUpdate({ bool onlyChanges: false, bool noGeometry: false }) {
```
Should this have an early-out if there's no one looking at semantics?

Hixie replied:
Yeah, good call. Until someone looks at semantics, everyone's dirty anyway. I originally figured that this would mean this function automatically "early-outed" in that situation but since then it's gotten a bit more involved...

abarth commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -952,3 +952,52 @@ class _TabBarViewState extends PageableListState<TabBarView> implements TabBarSe
   }
 }
+
+class TabPageSelector<T> extends StatelessComponent {
```
Where is this class coming from?

Hixie replied:
This is the code that was moved from the file at the very top of this patch.

abarth commented on examples/material_gallery/lib/main.dart:
```diff
@@ -193,10 +193,10 @@ class GalleryHome extends StatelessComponent {
 
 void main() {
-  runApp(new MaterialApp(
+  runApp(new SemanticsDebugger(child: new MaterialApp(
```
You probably don't want to land this

Hixie replied:
fixed and filed https://github.com/flutter/flutter/issues/1384 to reexpose it

abarth commented on examples/material_gallery/lib/demo/page_selector_demo.dart:
```diff
@@ -2,49 +2,16 @@
 import 'package:flutter/material.dart';
 
 class PageSelectorDemo extends StatelessComponent {
-  Widget _buildTabIndicator(BuildContext context, String iconName) {
```
Maybe move this code in a separate patch?

Hixie replied:
The reason I did it in this patch is that I did the move in order to add the accessibility logic (it requires refactoring where the animation happens so that I have the builder outside all the dots instead of per-dot, and once I was rewriting the whole thing anyway, I figured I'd move it at the same time). But I can do the move without the accessibility logic, then just patch it in this patch if you prefer.

abarth commented:
Please add some more tests before landing.  LGTM.

In general, this looks great.  There are some things I hope we can improve, but we can do that in subsequent patches:

1. We're adding too much code to object.dart.  Hopefully we can find a way to factor the dependencies so that we can move most of this code out into a lower-level library.
2. The structure of the dirty bits in markNeedsSemanticUpdate doesn't seem ideal.  It's odd to keep a bit that indicates whether you've succeeded in dirtying all your children.  The usual approach is to set a bit on yourself that forces a walk of all your children in the flush phase.
3. Computing the transform and clip by walking upward is generally inefficient.
4. We should make sure we minimize overhead when no semantics sink is attached.

Hixie commented:
Latest patch is uploaded, including more tests.

I've introduced some bug in the stocks app whereby the overlay no longer works, but once I've fixed that I'll land this.
